internal/cmd/helm-operator/run/cmd.go

```diff
@@ -250,3 +265,36 @@ func exitIfUnsupported(options manager.Options) {
 		os.Exit(1)
 	}
 }
+
+func buildCacheFunc(ws []watches.Watch, sch *apimachruntime.Scheme) (cache.NewCacheFunc, error) {
+	selectorsByObject := cache.SelectorsByObject{}
+	chartNames := make([]string, 0, len(ws))
+	for _, w := range ws {
+		sch.AddKnownTypeWithName(w.GroupVersionKind, &unstructured.Unstructured{})
+
+		crObj := &unstructured.Unstructured{}
+		crObj.SetGroupVersionKind(w.GroupVersionKind)
+		sel, err := metav1.LabelSelectorAsSelector(&w.Selector)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse watch selector for %s: %v", w.GroupVersionKind, err)
+		}
+		selectorsByObject[crObj] = cache.ObjectSelector{Label: sel}
+
+		chrt, err := loader.LoadDir(w.ChartDir)
+		if err != nil {
+			return nil, fmt.Errorf("unable to load chart for %s: %v", w.GroupVersionKind, err)
+		}
+		chartNames = append(chartNames, chrt.Name())
+	}
+
+	req, err := labels.NewRequirement("helm.sdk.operatorframework.io/chart", selection.In, chartNames)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create label requirement for cache default selector: %v", err)
+	}
+	defaultSelector := labels.NewSelector().Add(*req)
+
+	return cache.BuilderWithOptions(cache.Options{
+		SelectorsByObject: selectorsByObject,
```

Comment:
Just to understand this right, we cache objects based on labels from all the charts in `watches.yaml` only when - a single namespace, or all namespaces are watched. When multiple namespaces are watched, this gets overridden here (https://github.com/operator-framework/operator-sdk/blob/d21ed6499ebfc8ecdb4508e1c2a2a0cfd2a151f3/internal/cmd/helm-operator/run/cmd.go#L165) and we watch all the objects from the specified multiple namespaces. Is this intentional?

Seems like it shouldn't be the case, and if we are adding an object selector for caching then single or multiple namespaces shouldn't matter. Or am I missing something?

Reply:
I'm not sure on how urgent it is to have this in. We can either track this through an issue and make sure we get this fixed as a follow up after the bump or block the PR on #6514

Reply:
I think I managed to account for multiple namespaces without having to bump controller-runtime. PTAL!